// create-medusa-app/dist/utils/project-creator/creator.js
import ora from "ora";
import path from "path";
import createAbortController from "../create-abort-controller.js";
import ProcessManager from "../process-manager.js";
// Base class for common project functionality
export class BaseProjectCreator {
    options;
    args;
    spinner;
    processManager;
    abortController;
    factBoxOptions;
    projectName;
    projectPath;
    isProjectCreated = false;
    printedMessage = false;
    constructor(projectName, options, args) {
        this.options = options;
        this.args = args;
        this.spinner = ora();
        this.processManager = new ProcessManager();
        this.abortController = createAbortController(this.processManager);
        this.projectName = projectName;
        const basePath = typeof options.directoryPath === "string" ? options.directoryPath : "";
        this.projectPath = path.join(basePath, projectName);
        this.factBoxOptions = {
            interval: null,
            spinner: this.spinner,
            processManager: this.processManager,
            message: "",
            title: "",
            verbose: options.verbose || false,
        };
    }
    getProjectPath(projectName) {
        return path.join(this.options.directoryPath ?? "", projectName);
    }
}
